### PR TITLE
[CLANG][AArch64] Add the  modal 8 bit floating-point scalar type

### PR DESCRIPTION
**Pull request by CarolineConcatto**
ARM ACLE PR#323[1] adds new modal types for 8-bit floating point intrinsic.

From the PR#323:
```
ACLE defines the `__mfp8` type, which can be used for the E5M2 and E4M3
8-bit floating-point formats. It is a storage and interchange only type
with no arithmetic operations other than intrinsic calls.
````

The type should be an opaque type and its format in undefined in Clang. Only defined in the backend by a status/format register, for AArch64 the FPMR.

This patch is an attempt to the add the mfloat8_t scalar type. It has a parser and codegen for the new scalar type.

The patch it is lowering to and 8bit unsigned as it has no format. But maybe we should add another opaque type.

[1]  https://github.com/ARM-software/acle/pull/323

### CODE REVIEW

**Comment by CarolineConcatto** [comment]
Just in I don't believe this patch is ready, but I would like to share so I could get some pointers of what is missing.
I  also have  question about implementation:
Should we convert to a)unsigned 8 bits correspondent in CodeGenTypes.cpp or b)should we create another opaque  in LLVM  or c) something else?
In the implementation should we use  UNSIGNED_TYPE(Fpm8, Fpm8Ty) in BuiltinTypes.def or needs to adds OPAQUE?

File: clang/lib/Sema/SemaCast.cpp [hunk 1]
```diff
@@ -3197,6 +3197,13 @@ void CastOperation::CheckCStyleCast() {
     }
   }
 
+  if ((DestType->isMFloat8Type() && !SrcType->isMFloat8Type()) ||
+      (!DestType->isMFloat8Type() && SrcType->isMFloat8Type())) {
```
Comment (paulwalker-arm) [inline clang/lib/Sema/SemaCast.cpp:3201]: `DestType->isMFloat8Type() != SrcType->isMFloat8Type()` ?

File: clang/include/clang/AST/ASTContext.h [hunk 2]
```diff
@@ -1139,6 +1139,7 @@ class ASTContext : public RefCountedBase<ASTContext> {
   CanQualType SatShortFractTy, SatFractTy, SatLongFractTy;
   CanQualType SatUnsignedShortFractTy, SatUnsignedFractTy,
       SatUnsignedLongFractTy;
+  CanQualType MFloat8Ty;
```
Comment (paulwalker-arm) [inline clang/include/clang/AST/ASTContext.h:1142]: For this and all related "in llvm code" naming, do you mind using the `ArmMFloat8` prefix instead of `MFloat8`?  I ask because BFloat16 has an architecture independent meaning whereas this new type is specific to Arm and its related builtins.

File: clang/include/clang/AST/BuiltinTypes.def [hunk 3]
```diff
@@ -221,6 +221,10 @@ FLOATING_TYPE(Float128, Float128Ty)
 // '__ibm128'
 FLOATING_TYPE(Ibm128, Ibm128Ty)
 
+
+// '__fpm8'
```
Comment (paulwalker-arm) [inline clang/include/clang/AST/BuiltinTypes.def:225]: __mfp8

**Comment by paulwalker-arm** [comment]
Just to add the commit message needs updating to reflect the updated specification.

File: clang/include/clang/AST/ASTContext.h (hunk 2 above)
Comment (CarolineConcatto) [reply clang/include/clang/AST/ASTContext.h:1142]: I am not sure. The closest I see to a target dependent type are some openCL types
 OCLClkEventTy like OCLQueueTy.
I will change to ArmMFloat8 as you suggested.

**Review by paulwalker-arm — COMMENTED** [review]
Sorry, my comments are a bit all over the place, but I'm concerned we've not defined enough (and/or I need to understand what it means to be a storage only type) for me to confidently accept this PR.

If possible, my suggestion is to detach the scalar type support from the vector type support because I think the vector side of things is more complete from a specification point of view.

File: clang/include/clang/Sema/DeclSpec.h [hunk 4]
```diff
@@ -325,6 +325,8 @@ class DeclSpec {
 #define HLSL_INTANGIBLE_TYPE(Name, Id, SingletonId)                            \
   static const TST TST_##Name = clang::TST_##Name;
 #include "clang/Basic/HLSLIntangibleTypes.def"
+  // AARCH64_OPAQUE_TYPE
+  static const TST TST_ArmMFloat8_t = clang::TST_ArmMFloat8_t;
```
[the remark this answers is not in the document]
Comment (CarolineConcatto) [reply clang/include/clang/Sema/DeclSpec.h:329]: So, I was looking where this is used, and I can see it is used in many places for  parser and sema.
If **DeclSpec::TST_ArmMFloat8_t**  is removed from this places  the compiler complains it does not know the type mfloat8_t(  "unknown type name 'mfloat8_t';") in the header  arm_mfp8.h

[the remark this answers is not in the document]
Comment (CarolineConcatto) [reply clang/include/clang/Sema/DeclSpec.h]: It looks this is used in sema and to parse.

File: clang/lib/CodeGen/CGDebugInfo.cpp [hunk 5]
```diff
@@ -782,6 +782,13 @@ llvm::DIType *CGDebugInfo::CreateType(const BuiltinType *BT) {
 #define SVE_TYPE(Name, Id, SingletonId) case BuiltinType::Id:
 #include "clang/Basic/AArch64SVEACLETypes.def"
     {
+      if (BT->getKind() == BuiltinType::ArmMFloat8) {
+        Encoding = llvm::dwarf::DW_ATE_unsigned_char;
+        BTName = BT->getName(CGM.getLangOpts());
+        // Bit size and offset of the type.
+        uint64_t Size = CGM.getContext().getTypeSize(BT);
+        return DBuilder.createBasicType(BTName, Size, Encoding);
+      }
```
[the remark this answers is not in the document]
Comment (CarolineConcatto) [reply clang/lib/CodeGen/CGDebugInfo.cpp:791]: The codgen for the MFloat8 is now a vector <1 x i8>.

File: clang/lib/AST/Type.cpp [hunk 6]
```diff
@@ -3437,8 +3445,8 @@ StringRef BuiltinType::getName(const PrintingPolicy &Policy) const {
   case Id: \
     return #ExtType;
 #include "clang/Basic/OpenCLExtensionTypes.def"
-#define SVE_TYPE(Name, Id, SingletonId) \
-  case Id: \
+#define SVE_TYPE(Name, Id, SingletonId)                                        \
+  case Id:                                                                     \
```
Comment (paulwalker-arm) [inline clang/lib/AST/Type.cpp:3449]: Presumably unintentional?

File: clang/lib/CodeGen/CodeGenTypes.cpp [hunk 7]
```diff
@@ -476,7 +476,6 @@ llvm::Type *CodeGenTypes::ConvertType(QualType T) {
                                     Context.getFloatTypeSemantics(T),
                                     /* UseNativeHalf = */ false);
       break;
-
```
Comment (paulwalker-arm) [inline clang/lib/CodeGen/CodeGenTypes.cpp:479]: Unnecessary whitespace change.

File: clang/lib/CodeGen/CodeGenTypes.cpp [hunk 8]
```diff
@@ -526,6 +526,11 @@ llvm::Type *CodeGenTypes::ConvertType(QualType T) {
       }
     case BuiltinType::SveCount:
       return llvm::TargetExtType::get(getLLVMContext(), "aarch64.svcount");
+    case BuiltinType::ArmMFloat8:
+      ResultType =
+          llvm::VectorType::get(llvm::Type::getInt8Ty(getLLVMContext()),
+                                llvm::ElementCount::getFixed(1));
+      break;
```
Comment (paulwalker-arm) [inline clang/lib/CodeGen/CodeGenTypes.cpp:533]: Not sure how good an idea this is but what about updating `getBuiltinVectorTypeInfo` so the code above can be used? I mean, with the new definition mfloat8_t is effectively a vector type of sorts.

File: clang/lib/Parse/ParseDecl.cpp [hunk 9]
```diff
@@ -5828,6 +5832,7 @@ bool Parser::isKnownToBeTypeSpecifier(const Token &Tok) const {
   case tok::kw__ExtInt:
   case tok::kw__BitInt:
   case tok::kw___bf16:
+  case tok::kw___mfp8:
```
Comment (paulwalker-arm) [inline clang/lib/Parse/ParseDecl.cpp:5835]: Do you know why this and the similar changes are needed?  A quick search yielded nothing similar for the SVE type names, so just wondering if we're missing some auto-generation somewhere.

File: clang/lib/AST/ASTContext.cpp [hunk 10]
```diff
@@ -4302,6 +4308,7 @@ ASTContext::getBuiltinVectorTypeInfo(const BuiltinType *Ty) const {
   case BuiltinType::Id:                                                        \
     return {BoolTy, llvm::ElementCount::getScalable(NumEls), NF};
 #define SVE_OPAQUE_TYPE(Name, MangledName, Id, SingletonId)
```
[the remark this answers is not in the document]
Comment (CarolineConcatto) [reply clang/lib/AST/ASTContext.cpp:4385]: I am not sure I understand why do you mean. I cannot remove SVE_OPAQUE_TYPE from here.

[the remark this answers is not in the document]
Comment (paulwalker-arm) [reply clang/lib/AST/ASTContext.cpp:4385]: This was a post rebase comment, so the affected code is no longer in this PR.

File: clang/include/clang/Basic/AArch64SVEACLETypes.def [hunk 11]
```diff
@@ -107,6 +107,15 @@
...
 
+#ifndef AARCH64_SCALAR_TYPE
+#define AARCH64_SCALAR_TYPE(Name, MangledName, Id, SingletonId) \
+  SVE_TYPE(Name, Id, SingletonId)
+#endif
+
+#ifndef AARCH64_SCALAR_TYPE_MFLOAT
+#define AARCH64_SCALAR_TYPE_MFLOAT(Name, MangledName, Id, SingletonId, NumEls, ElBits, NF) \
+  AARCH64_SCALAR_TYPE(Name, MangledName, Id, SingletonId)
+#endif
```
Comment (paulwalker-arm) [inline clang/include/clang/Basic/AArch64SVEACLETypes.def:118]: Hopefully this isn't a bum steer but now everything has settled down I think these are no longer necessary and you can use `AARCH64_VECTOR_TYPE_MFLOAT` to define the scalar type as well because its internal representation is still a vector.

Comment (CarolineConcatto) [reply clang/include/clang/Basic/AArch64SVEACLETypes.def:118]: Yes, I agree with you that we dont need a scalar type anymore, because the MFloat8_t will also be a vector.

**Review by paulwalker-arm — APPROVED** [review]

**Comment by DavidSpickett** [comment]
FYI Arm 32 bit is failing on a different test than most of these - https://lab.llvm.org/buildbot/#/builders/154/builds/6442

File: clang/test/AST/arm-mfp8.cpp [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+//CHECK-NEXT:  | | |-CXXCtorInitializer {{.*}} 'f3c' 'volatile __MFloat8_t'
+//CHECK-NEXT:  | | | `-ImplicitCastExpr {{.*}} '__mfp8':'__MFloat8_t' <LValueToRValue>
+//CHECK-NEXT:  | | |   `-DeclRefExpr {{.*}} '__mfp8':'__MFloat8_t' lvalue ParmVar {{.*}} 'arg' '__mfp8':'__MFloat8_t'
+//CHECK-NEXT:  | | `-CompoundStmt {{.*}}
+//CHECK-NEXT:  | |-CXXMethodDecl {{.*}} func1c '__mfp8 (__mfp8)' implicit-inline
+//CHECK-NEXT:  | | |-ParmVarDecl {{.*}} arg '__mfp8':'__MFloat8_t'
+//CHECK-NEXT:  | | `-CompoundStmt {{.*}}
+//CHECK-NEXT:  | |   `-ReturnStmt {{.*}}
+//CHECK-NEXT:  | |     `-ImplicitCastExpr {{.*}} '__mfp8':'__MFloat8_t' <LValueToRValue>
+//CHECK-NEXT:  | |       `-DeclRefExpr {{.*}} '__mfp8':'__MFloat8_t' lvalue ParmVar {{.*}}8 'arg' '__mfp8':'__MFloat8_t'
```
Comment (DavidSpickett) [inline clang/test/AST/arm-mfp8.cpp:72]: Perhaps this `}8` is a typo and it should be `{{.*}}` only? I will try it locally.

Comment (paulwalker-arm) [reply clang/test/AST/arm-mfp8.cpp:72]: It's perhaps because the test uses `aarch64-registered-target || arm-registered-target` for what's effectively an AArch64 specific test. From what I can see there are many such tests that are miss classified but I guess fp8 being strictly AArch64 specific means this is the first time we get an explicit failure?

Comment (DavidSpickett) [reply clang/test/AST/arm-mfp8.cpp:72]: The test passed with the 8 removed so I committed https://github.com/llvm/llvm-project/commit/81e536ec87a108d012cf9156a2c3fc672fb92155. As the number its checking doesn't seem to be significant. I can revert if that's going to cause other problems.

Comment (DavidSpickett) [reply clang/test/AST/arm-mfp8.cpp:72]: If I recall correctly, clang can always produce IR or AST in this case, for a given target triple regardless of the backends included. So the `aarch64-registered-target || arm-registered-target` is probably just so including `arm_neon.h` doesn't error?

If you actually tried to use the fp8 stuff on Arm 32, you'd have other problems I assume.
